internal/volumehelper/volume_policy_helper.go

```diff
@@ -0,0 +1,243 @@
+				return true, nil
+			}
+			return false, nil
+		}
+	}
+
+	// now if volumepolicy is not specified then just check for snapshotVolumes flag
+	if boolptr.IsSetToTrue(v.SnapshotVolumes) {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+func (v *VolumeHelperImpl) GetVolumesForFSBackup(pod *corev1api.Pod) ([]string, []string, error) {
+	// Check if there is a fs-backup/snapshot volume policy specified by the user, if yes then use the volume policy approach to
+	// get the list volumes for fs-backup else go via the legacy annotation based approach
+	var includedVolumes, optedOutVolumes, volsToProcessByLegacyApproach = make([]string, 0), make([]string, 0), make([]string, 0)
+	var err error
+
+	if v.VolumePolicy != nil {
+		// Get the list of volumes to back up using pod volume backup for the given pod matching fs-backup volume policy action
+		v.Logger.Infof("Volume Policy specified by the user, using volume policy approach to segregate pod volumes for fs-backup")
+		includedVolumes, optedOutVolumes, volsToProcessByLegacyApproach, err = v.GetVolumesMatchingFSBackupAction(pod, v.VolumePolicy)
+		if err != nil {
+			return includedVolumes, optedOutVolumes, err
+		}
+	}
+
+	// process legacy annotation based approach, this will be done for 2 scenarios:
+	// 1. volume policy not specified by the user
+	// 2. there are some volumes for which the volume policy approach did not get any supported matching actions
+	if v.VolumePolicy == nil || len(volsToProcessByLegacyApproach) > 0 {
+		// Get the list of volumes to back up using pod volume backup from the pod's annotations.
+		// We will also pass the list of volume that did not have any supported volume policy action matched in legacy approach so that
+		// those volumes get processed via legacy annotation based approach, this is a fallback option on annotation based legacy approach
+		v.Logger.Infof("fs-backup or snapshot Volume Policy not specified by the user, using legacy approach based on annotations")
+		includedVolumes, optedOutVolumes = pdvolumeutil.GetVolumesByPod(pod, v.DefaultVolumesToFsBackup, v.BackupExcludePVC, volsToProcessByLegacyApproach)
+	}
+	return includedVolumes, optedOutVolumes, nil
+}
+
+// GetVolumesMatchingFSBackupAction returns a list of volume names to backup for the provided pod having fs-backup volume policy action
+func (v *VolumeHelperImpl) GetVolumesMatchingFSBackupAction(pod *corev1api.Pod, volumePolicies *resourcepolicies.Policies) ([]string, []string, []string, error) {
+	FSBackupActionMatchingVols := []string{}
+	FSBackupNonActionMatchingVols := []string{}
+	NoActionMatchingVols := []string{}
+
+	volsToExclude := pdvolumeutil.GetVolumesToExclude(pod)
+	for _, vol := range pod.Spec.Volumes {
+		if !v.ShouldIncludeVolumeInBackup(vol, volsToExclude) {
+			continue
+		}
+		// don't backup volumes that are included in the exclude list.
+		if pdvolumeutil.Contains(volsToExclude, vol.Name) {
```

Comment:
Is the condition `pdvolumeutil.Contains` in this line of code duplicated with the same logic in `ShouldIncludeVolumeInBackup` function? or it could not execute Line 130 for it will always execute  L126 first if the matching condition?

Reply:
@shubham-pampattiwar please take a look before merging this PR

Reply:
Would it should be like below?
```golang
	for i, vol := range pod.Spec.Volumes {
		if !v.ShouldIncludeVolumeInBackup(vol, volsToExclude, backupExcludePVC) {
		       // don't backup volumes that are included in the exclude list.
		        if util.Contains(volsToExclude, vol.Name) {
			     FSBackupNonActionMatchingVols = append(FSBackupNonActionMatchingVols, vol.Name)
			     }
			continue
		}
		
		if vol.PersistentVolumeClaim != nil {}
```

Reply:
Actually I have removed the `volsToExclude` check from the volume policy processing logic as it is related to legacy annotations based approach.
To Clarify, Volume Policy existence will be checked first and processed first. The legacy annotations based fallback option will be used only in 2 cases:
- If no Volume Policy is specified by the user: [Ref code link from PR](https://github.com/vmware-tanzu/velero/pull/7664/files#diff-412f1e7719b8fbd177428c1cb421a4b1caef8e0c33d78f7b83f6dcbafae87053R114)
- If there are Volumes for which no Volume Policy action was found i.e. action was nil : [Ref code link from PR](https://github.com/vmware-tanzu/velero/pull/7664/files#diff-412f1e7719b8fbd177428c1cb421a4b1caef8e0c33d78f7b83f6dcbafae87053R102)
I have added comments in the code for the above process.